integrations/helm/releasesync/releasesync.go

```diff
@@ -0,0 +1,349 @@
+	relInfo := make(map[string][]ReleaseFhr)
+
+	for _, ns := range namespaces {
+		list, err := rs.getNSCustomResources(ns)
+		if err != nil {
+			rs.logger.Log("error", fmt.Errorf("Failure while retrieving FluxHelmReleases in namespace %s: %v", ns, err))
+			return nil, err
+		}
+		rf := []ReleaseFhr{}
+		for _, fhr := range list.Items {
+			relName := release.GetReleaseName(fhr)
+			rf = append(rf, ReleaseFhr{RelName: relName, Fhr: fhr})
+		}
+		if len(rf) > 0 {
+			relInfo[ns] = rf
+		}
+	}
+	return relInfo, nil
+}
+
+// GetCustomResourcesLastDeploy retrieves and stores last event timestamp for FluxHelmRelease resources (FHR)
+// (Tamara: The method is not currently used. I am leaving it here as it will come handy for display on the UI)
+//		output:
+//             map[namespace][FHR name] = int64
+func (rs *ReleaseChangeSync) GetCustomResourcesLastDeploy(namespaces []string) (map[string]map[string]int64, error) {
+	relEventsTime := make(map[string]map[string]int64)
+
+	for _, ns := range namespaces {
+		eventList, err := rs.getNSEvents(ns)
+		if err != nil {
+			return relEventsTime, err
+		}
+		fhrD := make(map[string]int64)
+		for _, e := range eventList.Items {
+			if e.InvolvedObject.Kind == CustomResourceKind {
+				secs := e.LastTimestamp.Unix()
+				fhrD[e.InvolvedObject.Name] = secs
+			}
+		}
+		relEventsTime[ns] = fhrD
+	}
+	return relEventsTime, nil
+}
+
+func (rs *ReleaseChangeSync) shouldUpgrade(currRel *hapi_release.Release, fhr ifv1.FluxHelmRelease) (bool, error) {
+	currVals := currRel.GetConfig().GetRaw()
+	currChart := currRel.GetChart().String()
+
+	// Get the desired release state
+	opts := chartrelease.InstallOptions{DryRun: true}
+	tempRelName := strings.Join([]string{currRel.GetName(), "temp"}, "-")
+	desRel, err := rs.release.Install(rs.release.Repo.ReleasesSync, tempRelName, fhr, "CREATE", opts)
+	if err != nil {
+		return false, err
+	}
+	desVals := desRel.GetConfig().GetRaw()
+	desChart := desRel.GetChart().String()
+
+	// compare values && Chart
+	if strings.Compare(currVals, desVals) != 0 {
+		rs.logger.Log("error", fmt.Sprintf("Release %s: values have diverged due to manual Chart release", currRel.GetName()))
+		return true, nil
+	}
+	if strings.Compare(currChart, desChart) != 0 {
+		rs.logger.Log("error", fmt.Sprintf("Release %s: Chart has diverged due to manual Chart release", currRel.GetName()))
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// existingReleasesToSync determines which Chart releases need to be deleted/upgraded
+// to bring the cluster to the desired state
+func (rs *ReleaseChangeSync) existingReleasesToSync(
+	currentReleases map[string]map[string]int64,
+	customResources map[string]map[string]ifv1.FluxHelmRelease,
+	relsToSync map[string][]chartRelease) error {
+
+	var chRels []chartRelease
+	for ns, nsRelsM := range currentReleases {
+		chRels = relsToSync[ns]
+		for relName := range nsRelsM {
+			fhr, ok := customResources[ns][relName]
```

Comment:
It looks like this would panic if there's a current release in a namespace that doesn't have a customer resource in it.

Reply:
A great catch. Fixed.